---
title: "250528_GenentechComparison"
author: "Irika Sinha"
date: "2025-05-26"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Purpose

The purpose of this script is to take the file with manually identified cryptic exon coordinates and use those coordinates to identify left and right junction PSIs in the different conditions for i3N. 

## Packages

```{r packages}
CRANpackages <- c("tidyverse", "rjson")
biocPackages <- c("GenomicRanges")

for(i in CRANpackages){
  if(!(i %in% rownames(installed.packages()))){
    install.packages(pkgs = i)
  }
  library(package = i, character.only = T, quietly = T)
}

for(i in biocPackages){
  if(!(i %in% rownames(installed.packages()))){
    BiocManager::install(pkgs = i, ask = F)
  }
  library(package = i, character.only = T, quietly = T)
}

rm(CRANpackages,biocPackages,i)

```

## DEG analysis
Nextflow RNA-seq pipeline used to generate DESeq2 objects.
Used DESeq2 vignette to write analysis script: https://www.bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

```{r DESeq2}
# Load data
cryptic_df = read_delim("data/250509_CE_table.csv")
gtech_ce <- read_delim("data/250526_GenentechCEs.csv") %>% filter(isCryptic == "TRUE")

df <- data.frame()

for(i in 1:nrow(gtech_ce)){
  df_g <- rjson::fromJSON(rjson::fromJSON(json_str = gtech_ce$spliceVariantJSON[i],simplify=TRUE))  %>% as.data.frame()
  df_g$niceVariantName = gtech_ce$niceVariantName[i]
  df_g$symbol = gtech_ce$symbol[i]
  df_g$spliceset = gtech_ce$spliceset[i]
  df_g$simpleType = gtech_ce$simpleType[i]
  df_g <- df_g %>% mutate(coordinates = paste0("chr",seqnames,":",start,"-",end))
  df_g <- df_g %>% relocate(symbol, .before=seqnames) %>% 
    relocate(coordinates, .before=seqnames) %>% 
    relocate(simpleType, .before=seqnames)
  df <- rbind(df, df_g)
  rm(i, df_g)
}

df_CE <- df %>% filter(type == "E") %>% arrange(symbol) %>% filter(!(coordinates %in% cryptic_df$`Cryptic Exon Location
(hg38)`)) 

# More sieving lol
rmRow <- c()
for(i in 1:nrow(df_CE)){
  remove = F
  rows <- which(cryptic_df$`Gene Symbol` == df_CE$symbol[i])
  if(length(rows) != 0){
    test1 = as.character(df_CE$start[i])
    test2 = as.character(df_CE$end[i])
    strTest = toString(cryptic_df$`Cryptic Exon Location
(hg38)`[rows])
      
    if(str_detect(pattern=test1,string=strTest)){
      remove = T
    } else if(str_detect(pattern=test2,string=strTest)){
      remove = T
    }
    
    if(remove){
      rmRow <- append(rmRow,i)
    }
  }
}

df_CE <- df_CE[-rmRow,]

rm(i, remove, rmRow, rows,strTest, test1, test2)




```